File: fushi/test/pages/popup_pitch_transcriptions_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

/// TODO-688: the lookup popup must RENDER the IPA `transcriptions` that
/// TODO-687 block3 wired into the popup JSON.
///
/// 背景：每个 pitch group 在 popup JSON 里带 `transcriptions` 字符串数组（如
/// `['neꜜko']`），普通音高词典为空，Yomitan `ipa`-mode 词典非空。TODO-687 已把数据
/// 贯通到 `popup.js` 接收的 entry，但 `popup.js` 之前没渲染它，用户看不到 IPA。
/// 本任务在 pitch group 内把每条 transcription 渲染成 `[ipa]` 小标签。
///
/// 两层守护：
/// ① 行为级——用 Node 真执行 popup.js 的 `createPitchSection`，断言：普通词典
///    （空 transcriptions）不渲染 transcriptions 列表；含 transcriptions 的词典
///    把 `[neꜜko]` 等渲染出来；纯 IPA 词典（pitchPositions 为空）在去重分支下
///    仍保留并渲染 transcriptions（不被空位置守卫丢掉）。无 node 时 skip。
/// ② 源码级——静态扫描 popup.js，保证 transcriptions 渲染路径在位（即便无 node
///    也能在 CI 守住回归），且非恒真——断言具体的渲染 helper、调用点、去重分支守卫。
void main() {
  test(
    'popup renders IPA transcriptions in the pitch group (executes createPitchSection via node)',
    () async {
      final String? nodeExe = _resolveNode();
      if (nodeExe == null) {
        markTestSkipped(
            'node not found on PATH; skipping JS behavior execution');
        return;
      }

      final File jsTest = File(
        'test/pages/popup_pitch_transcriptions_test.js',
      );
      expect(
        jsTest.existsSync(),
        isTrue,
        reason: 'behavior harness ${jsTest.path} must exist',
      );

      final ProcessResult result = await Process.run(
        nodeExe,
        <String>[jsTest.path],
        workingDirectory: Directory.current.path,
      );

      expect(
        result.exitCode,
        0,
        reason: 'popup IPA transcriptions JS behavior test failed.\n'
            'stdout:\n${result.stdout}\nstderr:\n${result.stderr}',
      );
      expect(
        result.stdout.toString(),
        contains('all assertions passed'),
        reason: 'behavior harness must reach its success marker',
      );
    },
  );

  test('popup.js renders pitch-group transcriptions as IPA tags', () {
    final String js = File('assets/popup/popup.js').readAsStringSync();

    // The dedicated transcription builder must exist and emit a `[ipa]` tag.
    expect(
      js,
      contains('function createTranscriptionsHtml(transcriptions)'),
      reason: 'a transcriptions render helper must exist',
    );
    expect(
      js,
      contains('textContent: `[\${ipa}]`'),
      reason: 'each transcription must render as an `[ipa]` tag',
    );

    // createPitchGroup must actually call the builder and append it — non-tautological:
    // the builder existing alone does nothing if never invoked.
    final int builderDef = js.indexOf('function createTranscriptionsHtml');
    final int groupDef = js.indexOf('function createPitchGroup');
    expect(groupDef, greaterThanOrEqualTo(0));
    final int call = js.indexOf(
        'createTranscriptionsHtml(pitchData.transcriptions)', groupDef);
    expect(
      call,
      greaterThan(groupDef),
      reason:
          'createPitchGroup must build transcriptions from its pitch group data',
    );
    expect(builderDef, greaterThanOrEqualTo(0));
  });

  test('popup.js dedup branch keeps IPA-only pitch groups (empty positions)',
      () {
    final String js = File('assets/popup/popup.js').readAsStringSync();

    // The deduplicate-pitch-accents branch must not drop a group that has no
    // unique pitch positions but DOES have transcriptions, and it must forward
    // the transcriptions into createPitchGroup.
    final int dedup = js.indexOf('if (window.deduplicatePitchAccents)');
    expect(dedup, greaterThanOrEqualTo(0));
    final int hasTranscriptions =
        js.indexOf('pitch.transcriptions?.length', dedup);
    expect(
      hasTranscriptions,
      greaterThan(dedup),
      reason:
          'the dedup branch must keep IPA-only groups (transcriptions guard)',
    );
    final int forwarded =
        js.indexOf('transcriptions: pitch.transcriptions', dedup);
    expect(
      forwarded,
      greaterThan(dedup),
      reason:
          'the dedup branch must forward transcriptions into createPitchGroup',
    );
  });
}

/// Resolve a usable `node` executable, returning null when none is on PATH.
String? _resolveNode() {
  final List<String> candidates =
      Platform.isWindows ? <String>['node.exe', 'node'] : <String>['node'];
  for (final String name in candidates) {
    try {
      final ProcessResult probe = Process.runSync(name, <String>['--version']);
      if (probe.exitCode == 0) {
        return name;
      }
    } on ProcessException {
      // Not found; try next candidate.
    }
  }
  return null;
}

File: fushi/test/pages/popup_pitch_transcriptions_test.js
// TODO-688 behavior test: the popup must RENDER the IPA `transcriptions` that
// TODO-687 block3 wired into the popup JSON. Each pitch GROUP carries a
// `transcriptions` string array (e.g. ['neꜜko']) — empty for plain pitch-accent
// dicts, populated for Yomitan `ipa`-mode dicts. popup.js renders them inside the
// pitch group as `[ipa]` tags.
//
// This EXECUTES the real popup.js against a minimal fake DOM and drives the real
// `createPitchSection` builder, then walks the produced element tree for the
// transcription text. Reverting the render (dropping the createTranscriptionsHtml
// call, or the empty-pitchPositions guard in createPitchSection) turns this red.
//
// Cases covered:
//   1. plain pitch dict (transcriptions: []) renders NO transcription text.
//   2. dict with both pitch positions AND transcriptions renders both.
//   3. IPA-only dict (pitchPositions: [], transcriptions: [...]) — the group is
//      kept under deduplicatePitchAccents and the transcriptions still render.
//
// Run: node fushi/test/pages/popup_pitch_transcriptions_test.js
// (also driven from popup_pitch_transcriptions_test.dart so it executes inside
//  `flutter test`).

const assert = require('assert');
const {
  loadPopup,
  collectByClass,
  collectText,
} = require('./_popup_dom_host.js');


// Collect the className of every element node in the tree.
function collectClasses(node, acc) {
  acc = acc || [];
  if (!node) return acc;
  if (node.nodeType !== 3 && node.className) acc.push(node.className);
  const kids = node.children || node.childNodes || [];
  for (const k of kids) collectClasses(k, acc);
  return acc;
}

(function run() {
  // Case 1: plain pitch dict — transcriptions empty → no transcription markup.
  {
    const sb = loadPopup();
    const section = sb.window.__test.createPitchSection(
      [{ dictionary: 'NHK', pitchPositions: [2], transcriptions: [] }], 'ねこ');
    assert.ok(section, 'plain pitch dict must still render a pitch section');
    const classes = collectClasses(section);
    assert.ok(!classes.includes('pitch-transcriptions'),
      'a plain pitch dict (empty transcriptions) must NOT render a transcriptions list; got '
        + JSON.stringify(classes));
  }

  // Case 2: pitch positions AND transcriptions both present → both render.
  {
    const sb = loadPopup();
    const section = sb.window.__test.createPitchSection(
      [{ dictionary: 'IPA', pitchPositions: [1], transcriptions: ['neꜜko', 'neko'] }], 'ねこ');
    assert.ok(section, 'mixed dict must render a pitch section');
    const text = collectText(section);
    assert.ok(text.includes('[neꜜko]'),
      'transcription [neꜜko] must be rendered; got ' + JSON.stringify(text));
    assert.ok(text.includes('[neko]'),
      'all transcriptions must be rendered; got ' + JSON.stringify(text));
    assert.ok(text.includes('[1]'),
      'the pitch position must still render alongside transcriptions; got ' + JSON.stringify(text));
    const classes = collectClasses(section);
    assert.ok(classes.includes('pitch-transcriptions'),
      'a transcriptions list element must exist; got ' + JSON.stringify(classes));
  }

  // Case 3: IPA-only dict (no pitch positions) under deduplicatePitchAccents.
  // The group has no unique pitch positions, so the dedup branch would drop it
  // unless transcriptions keep it alive.
  {
    const sb = loadPopup();
    sb.window.deduplicatePitchAccents = true;
    const section = sb.window.__test.createPitchSection(
      [{ dictionary: 'IPA-only', pitchPositions: [], transcriptions: ['tabeꜜɾɯ'] }], 'たべる');
    assert.ok(section, 'IPA-only dict must render a pitch section');
    const text = collectText(section);
    assert.ok(text.includes('[tabeꜜɾɯ]'),
      'IPA-only transcription must NOT be dropped by the dedup empty-position guard; got '
        + JSON.stringify(text));
  }

  console.log('popup_pitch_transcriptions_test.js: all assertions passed');
})();
